onnxruntime/core/session/inference_session_utils.cc

```diff
@@ -0,0 +1,281 @@
+    }
+  }
+
+  if (!session_options_found) {
+    LOGS(logger, INFO) << "Did not find session options in the model file to be used while running the model";
+    return Status::OK();
+  }
+
+  auto& session_options_from_model = json_obj.at("session_options");
+
+  // TODO: Support all valid session options
+  // Only the following config options from the json will be supported in this version
+  // Any other option if part of the json (even if valid session option) will be ignored
+
+  for (auto it = session_options_from_model.cbegin(); it != session_options_from_model.cend(); ++it) {
+    auto& key = it.key();
+    auto& value = it.value();
+
+    if (key == "intra_op_num_threads ") {
+      if (!value.is_number_integer()) {
+        return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
+                               "intra_op_num_threads option in the model file must be an integer");
+      }
+
+      auto status = set_intra_op_num_threads(session_options, it.value().get<int>(), logger);
+
+      if (!status.IsOK()) {
+        return status;
+      }
+
+    } else if (key == "inter_op_num_threads") {
+      if (!value.is_number_integer()) {
+        return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
+                               "inter_op_num_threads option in the model file must be an integer");
+      }
+
+      auto status = set_inter_op_num_threads(session_options, it.value().get<int>(), logger);
+
+      if (!status.IsOK()) {
+        return status;
+      }
+    } else if (key == "execution_mode") {
+      if (!value.is_number_integer()) {
+        return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
+                               "execution_mode option in the model file must be an integer");
+      }
+
+      auto status = set_execution_mode(session_options, it.value().get<int>(), logger);
+
+      if (!status.IsOK()) {
+        return status;
+      }
+    } else if (key == "graph_optimization_level") {
+      if (!value.is_number_integer()) {
+        return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
+                               "graph_optimization_level option in the model file must be an integer");
+      }
+
+      auto status = set_graph_optimization_level(session_options, it.value().get<int>(), logger);
+
+      if (!status.IsOK()) {
+        return status;
+      }
+    } else if (key == "enable_profiling") {
+      if (!value.is_number_integer()) {
+        return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
+                               "enable_profiling option in the model file must be an integer");
+      }
+
+      auto status = set_enable_profiling(session_options, it.value().get<int>(), logger);
+
+      if (!status.IsOK()) {
+        return status;
+      }
+    } else {
+      LOGS(logger, INFO) << "Ignoring unsupported session option in ORT config: " << key;
+    }
+  }
+
+  return Status::OK();
+}
+
+Status parse_run_options_from_model_proto(const ONNX_NAMESPACE::ModelProto& /*model_proto*/,
```

Comment:
Can we parse the json blob only once? It look like every parse_* function will call parse_json_and_search_for_key and search for the appropriate key.

Reply:
It shouldn't matter for now as RunOptions is not supported, but yeah, it can be parsed just once. The impact shouldn't be tremendous given that the payload won't be super heavy. I ll refactor it to make the parsing happen only once though.

Reply:
Fixed. Parsing will only happen once.